engine/client/voice.c

```diff
@@ -628,67 +1167,53 @@ Initialize the voice subsystem
 */
 qboolean Voice_Init( const char *pszCodecName, int quality, qboolean preinit )
 {
-	if( Q_strcmp( pszCodecName, VOICE_OPUS_CUSTOM_CODEC ))
-	{
-		if( COM_CheckStringEmpty( pszCodecName ))
-			Con_Printf( S_ERROR "Server requested unsupported codec: %s\n", pszCodecName );
-
-		// reset saved codec name, we won't enable voice for this connection
-		voice_codec_init[0] = 0;
-		voice_quality_init = 0;
-		return false;
-	}
-
-	Q_strncpy( voice_codec_init, pszCodecName, sizeof( voice_codec_init ));
-	voice_quality_init = quality;
+	Q_strncpy( voice.codec, pszCodecName, sizeof( voice.codec ));
+	voice.quality = quality;
 
 	if( !voice_enable.value )
 		return false;
 
-	// reinitialize only if codec parameters are different
-	if( Q_strcmp( pszCodecName, voice.codec ) || voice.quality != quality )
-	{
-		Voice_Shutdown();
-		voice.autogain.block_size = 128;
+	if( preinit )
+		return true;
 
-		if( !Voice_InitCustomMode( ))
+	if( Voice_IsGoldSrcMode( pszCodecName ) )
+	{
+		if( !Voice_InitGoldSrcMode( quality ) )
 		{
-			// no reason to init encoder and open audio device
-			// if we can't hear other players
 			Voice_Shutdown();
 			return false;
 		}
-
-		// we can hear others players, so it's fine to fail now
-		voice.initialized = true;
-		Q_strncpy( voice.codec, pszCodecName, sizeof( voice.codec ));
-
-		if( !Voice_InitOpusEncoder( quality ))
-		{
-			Con_Printf( S_WARN "Other players will not be able to hear you.\n" );
-			return false;
-		}
-
-		voice.quality = quality;
 	}
-
-	if( !preinit )
+	else if( Voice_IsOpusCustomMode( pszCodecName ) )
 	{
-		Voice_ShutdownOpusDecoder();
-		if( !Voice_InitOpusDecoder())
+		if( !Voice_InitOpusCustomMode( quality ) )
 		{
-			// no reason to init encoder and open audio device
-			// if we can't hear other players
-			Con_Printf( S_ERROR "Can't create decoders, voice chat is disabled.\n" );
 			Voice_Shutdown();
 			return false;
 		}
+	}
+	else
+	{
+		// unsupported codec
+		Voice_Shutdown();
+		voice.goldsrc = false;
```

Comment:
Why? If the voice hasn't been initialized, cleaning up the whole structure is much more fool-proof.